version: "3.9"

services:
  redis:
    image: redis:alpine
    container_name: redis_broker
    networks:
      - idrl-network
  
  db:
    image: postgres
    container_name: fpv_db
    environment:
        POSTGRES_DB: fpv_db
        POSTGRES_USER: postgres
        POSTGRES_PASSWORD: postgres
        PGUSER: postgres # Added because try connection with root user
    networks:
      - idrl-network
    ports: 
      - "5432:5432"
    volumes: 
      - postgres-data:/var/lib/postgresql/data
    healthcheck:
      test: ["CMD-SHELL", "pg_isready"]
      interval: 5s
      timeout: 5s
      retries: 5

  user:
    container_name: user
    build: ./user
    environment:
        DB_URL: postgresql://postgres:postgres@db:5432/postgres
    volumes:
      - ./user:/app
      - ./videos:/app/videos
    networks:
      - idrl-network
    depends_on: 
       db:
        condition: service_healthy

  worker:
    container_name: worker
    build: ./worker
    environment:
      - FLASK_APP=app.py
      - FLASK_DEBUG=1
    volumes:
      - ./worker:/app
      - ./videos:/app/videos
    networks:
      - idrl-network
    depends_on:
      - db
      - redis
      
  api-gateway:
    image: nginx:latest
    container_name: api-gateway
    ports:
      - 81:80
    volumes:
      - ./nginx/nginx-proxy.conf:/etc/nginx/conf.d/default.conf
    depends_on:
      - user
    networks:
      - idrl-network

networks:
  idrl-network:
    driver: bridge

volumes: 
  postgres-data: 
    driver: local
  videos:
    driver: local
